### PR TITLE
Remove dependency on OpenCL/CUDA SDK

### PR DESCRIPTION
**Pull request by dsharletg**
This PR allows one to build programs with Halide that use OpenCL/CUDA without having the OpenCL/CUDA SDKs installed. It does this by dynamically loading OpenCL/CUDA at runtime.

In particular, this makes it much easier to use GPU enabled schedules in Matlab, which previously was difficult because mex_halide.m would need to support linking to OpenCL/CUDA libraries.

This needs to be tested on every platform before being merged. I've currently tested it on OSX+OpenCL. Linux+OpenCL and Linux+CUDA both appear to work, but my CUDA SDK install is busted somehow, so the acquire_release test doesn't build. This is the one test/app that still needs the CUDA/OpenCL SDK to build.

### CODE REVIEW

File: src/runtime/cuda.cpp [hunk 1]
```diff
@@ -4,8 +4,58 @@
 #include "mini_cuda.h"
 #include "cuda_opencl_shared.h"
 
+#define INLINE inline __attribute__((always_inline))
+
 namespace Halide { namespace Runtime { namespace Internal { namespace Cuda {
 
+// Define the function poiners for the CUDA API.
```
Comment (abadams) [inline src/runtime/cuda.cpp]: poiners -> pointers

File: src/runtime/runtime_internal.h [hunk 2]
```diff
@@ -86,6 +87,17 @@ WEAK char *halide_int64_to_string(char *dst, char *end, int64_t arg, int digits)
 WEAK char *halide_uint64_to_string(char *dst, char *end, uint64_t arg, int digits);
 WEAK char *halide_pointer_to_string(char *dst, char *end, const void *arg);
 
+// Search the current process for a symbol with the given name.
+WEAK void *halide_get_symbol(const char *name);
+// Platform specific implementations of dlopen/dlsym/dlclose.
+WEAK void *halide_load_library(const char *name);
+WEAK void *halide_get_library_symbol(void *lib, const char *name);
+WEAK void halide_free_library(void *lib);
```
Comment (abadams) [inline src/runtime/runtime_internal.h]: free doesn't seem like the right verb here. How about close or release?

Also, this function never seems to be called. I guess that makes sense for cuda/opencl. Will we ever want to close a library?

**Comment by abadams** [comment]
I kicked off the windows tests. Will report back.

File: src/runtime/runtime_internal.h (hunk 2 above)
Comment (dsharletg) [inline src/runtime/runtime_internal.h]: Good point, I'll just remove it for now. BTW, it's named free_library out of habit: on windows, the functions are LoadLibrary and FreeLibrary.

**Comment by abadams** [comment]
I get unresolved external symbol "LoadLibrary" everywhere on windows.

**Comment by dsharletg** [comment]
I just pushed a fix for that, thanks to @aam for realizing the problem.

**Comment by abadams** [comment]
On windows I now get an error in test/correctness/argmax.cpp:

LLVM ERROR: Program used external function 'cuCtxPopCurrent_v2' which could not be resolved!

**Comment by dsharletg** [comment]
I'm running the final round of testing on Windows right now, I believe it will pass. I will report back again when its done. Please take a look.

**Comment by dsharletg** [comment]
Well, I ran into some unrelated issues and am now forced to rebuild LLVM to run the tests again. However, I'm basically certain that this branch is not introducing any bugs on windows at this point.

**Comment by dsharletg** [comment]
I have been chasing my tail all day on this. There's a test failing (gpu_sum_scan with opencl), but it's also failing on master. We updated the GPU driver on this machine recently, and apparently we have something behaving differently. It might be our bug that we didn't reveal until updating the driver, or it might be a problem with the driver. Either way though, I'm ready to say that this branch is working and ready to be merged, and I'll continue looking at the failure on master.

**Comment by aam** [comment]
LGTM

This test(gpu_sum_scan with opencl) does pass on my Windows VS2015 machine. I also built and ran apps/local_laplacian on Windows to confirm that nvidia libraries are loaded correctly for cuda and opencl targets.

**Pull request merged**